addons/binding/org.openhab.binding.icloud/src/main/java/org/openhab/binding/icloud/handler/DeviceHandler.java

```diff
@@ -0,0 +1,171 @@
+    private final Logger logger = LoggerFactory.getLogger(DeviceHandler.class);
+    private String deviceId;
+    private BridgeHandler bridge;
+    private LocationProvider locationProvider;
+
+    public DeviceHandler(@NonNull Thing thing, LocationProvider locationProvider) {
+        super(thing);
+        this.locationProvider = locationProvider;
+    }
+
+    public void update(ArrayList<Content> deviceInformationList) {
+        Content deviceInformationRecord = getDeviceInformationRecord(deviceInformationList);
+        if (deviceInformationRecord != null) {
+            deviceId = deviceInformationRecord.getId();
+            updateStatus(ThingStatus.ONLINE);
+
+            updateState(BATTERY_STATUS, new StringType(deviceInformationRecord.getBatteryStatus()));
+
+            Double batteryLevel = deviceInformationRecord.getBatteryLevel();
+            if (batteryLevel != Double.NaN) {
+                updateState(BATTERY_LEVEL, new DecimalType(deviceInformationRecord.getBatteryLevel() * 100));
+            }
+
+            if (deviceInformationRecord.getLocation() != null) {
+                updateLocationRelatedStates(deviceInformationRecord);
+            }
+            updateState(DEVICE_NAME, new StringType(deviceInformationRecord.getName()));
+        } else {
+            updateStatus(ThingStatus.OFFLINE);
+        }
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing iCloud device handler.");
+        initializeThing((getBridge() == null) ? null : getBridge().getStatus());
+    }
+
+    @Override
+    public void handleCommand(@NonNull ChannelUID channelUID, Command command) {
+        String channelId = channelUID.getId();
+        if (channelId.equals(FIND_MY_PHONE)) {
+            if (command == OnOffType.ON) {
+                bridge.findMyDevice(deviceId);
+                updateState(FIND_MY_PHONE, OnOffType.OFF);
+            }
+        }
+    }
+
+    @Override
+    public void dispose() {
+        bridge.unregisterDevice(this);
+        super.dispose();
+    }
+
+    private void updateLocationRelatedStates(Content deviceInformationRecord) {
+        DecimalType latitude = new DecimalType(deviceInformationRecord.getLocation().getLatitude());
+        DecimalType longitude = new DecimalType(deviceInformationRecord.getLocation().getLongitude());
+        DecimalType accuracy = new DecimalType(deviceInformationRecord.getLocation().getHorizontalAccuracy());
+
+        PointType location = new PointType(latitude, longitude);
+
+        updateState(LOCATION, location);
+        updateState(LOCATION_ACCURACY, accuracy);
+        updateState(LOCATION_LASTUPDATE, getLastLocationUpdateDateTimeState(deviceInformationRecord));
+
+        if (locationProvider != null) {
+            PointType homeLocation = locationProvider.getLocation();
+            if (homeLocation != null) {
+                DecimalType distanceFromHome = homeLocation.distanceFrom(location);
```

Comment:
Hm, looking at this, I feel that this channel as well is not a feature of the binding as this information is not provided from the remote service. This calculation can be done in a rule for ANY location item and we should imho address it together with the address lookup feature, if you don't mind.

Reply:
I have removed the channel "DistanceFromHome".